### ISSUE TITLE
Unexpected behavior when printing the class instance object

### ISSUE THREAD

**Issue by oscardelama** [wch/R6#37]
Hi Winston.

I have a R6 class with an active binding named `cov.df`, which -of course- triggers some logic when that *variable* is read. For example, when this *variable* is read "too early" with respect to the expected usage of the class, the *variable* is empty, and its reading causes a friendly warning, giving to the user a hint about why the *variable* is empty. Something like this:

```R
cov.df = function(value) {
  #  <some previous code>
  if (nrow(private$.cov.df) == 0)
    warning('There is no "cov" information. You should probably run the digest() function before.')
  return(private$.cov.df)
  # <some following code>
}
```
So far, so good. However, when I "print" the class instance object I get that friendly warning as if the variable has been read, but if I am just printing the class components, such reading is neither required nor expected:

```R
> vo <- vvm$new()
> vo
#> <vvm>
#>   Public:
#>     <other class members>
#>     cov.df: active binding
#>     <more class members>
#>
#> Warning message:
#> In (function (value)  :
#>   There is no "cov" information. You should probably run the digest() function before.
```

In this case, the triggered behavior when *printing* the instance is harmless and I am commenting out those warnings while looking for a more suitable solution. Nevertheless, I am afraid those variable readings might have a negative -unexpected- impact, changing the state of the class, while it is trying to produce the requested (read) value.

Is there a way to circumvent these readings during the printing? i.e. Is there any flag letting me know the reading is made during a printing?

**Comment by gaborcsardi** [comment]
I think you can define your own print method as a workaround.

**Comment by oscardelama** [comment]
@wch: You say: 
> The print.R6 method will try to print out everything, and if you want different behavior 
> you'll have to implement your own print method.

I think you haven't understood entirely the issue. It is OK if the print.R6 method "prints everything". That is not the issue. The issue is that in the process, the `print.R6` method **reads the content** of the active binding variables, which is really not required, because it (the print.R6 method) only print their names, not their contents.

An active binding may be linked to a *variable* representing a "synthetic" value, which is built by request: when the *variable* (active binding) is read. So, the reading of those variables during the printing process, unexpectedly triggers the execution of code with different effects, from warning or error messages to changing the state of the instance object. This is not a trivial issue.

I think that fixing this issue for my classes might have, for me, almost the same cost than fixing it in the original print.r6 method. I will take a look to see if I can propose a pull request.

**Comment by gaborcsardi** [comment]
Oh, I see, sorry, my bad.

I guess you need to rewrite this part: https://github.com/wch/R6/blob/618a64302e62b2f6ade8035d085e7c935311f27c/R/print.R#L73
to only extract `x[[name]]` into `obj` if it is not an active binding. You just need to reorganize a couple of lines.

**Comment by oscardelama** [comment]
It's done.
The change was minimal. Basically it is just the change of sequence of two statements.
I have also added few lines of code to the test "Active bindings work" in `test-portable.r` to include this case and make sure this issue stays fixed. ;-)

**Comment by wch** [comment]
Ah, I see now. Thanks for submitting the PR.

**Comment by wch** [comment]
Fixed by #38.
